github/git_refs.go

```diff
@@ -58,13 +60,61 @@ func (s *GitService) GetRef(ctx context.Context, owner string, repo string, ref
 
 	r := new(Reference)
 	resp, err := s.client.Do(ctx, req, r)
-	if err != nil {
+	if _, ok := err.(*json.UnmarshalTypeError); ok {
+		// Multiple refs, means there wasn't an exact match.
+		return nil, resp, nil
+	} else if err != nil {
 		return nil, resp, err
 	}
 
 	return r, resp, nil
 }
 
+// GetRefs fetches a slice of Reference objects for a given Git ref.
+// If there is an exact match, only that ref is returned.
+// If there is no exact match, GitHub returns all refs that start with ref.
+// If returned error is nil, there will be at least 1 ref returned.
+// For example:
+//
+// 	"heads/featureA"  -> ["refs/heads/featureA"]                         // Exact match, single ref is returned.
+// 	"heads/feature"   -> ["refs/heads/featureA", "refs/heads/featureB"]  // All refs that start with ref.
+// 	"heads/notexist"  -> []                                              // Returns an error.
+//
+// GitHub API docs: https://developer.github.com/v3/git/refs/#get-a-reference
+func (s *GitService) GetRefs(ctx context.Context, owner string, repo string, ref string) ([]*Reference, *Response, error) {
+	ref = strings.TrimPrefix(ref, "refs/")
+	u := fmt.Sprintf("repos/%v/%v/git/refs/%v", owner, repo, ref)
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var rawJSON json.RawMessage
+	resp, err := s.client.Do(ctx, req, &rawJSON)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	// Prioritize the most common case: a single returned ref.
+	r := new(Reference)
+	singleUnmarshalError := json.Unmarshal(rawJSON, r)
+	if singleUnmarshalError == nil {
+		return []*Reference{r}, resp, nil
+	}
+
+	// Attempt to unmarshal multiple refs.
+	var rs []*Reference
+	multipleUnmarshalError := json.Unmarshal(rawJSON, &rs)
+	if multipleUnmarshalError == nil {
```

Comment:
@gmlewis I don't think this applies here, it's a special case. We don't know what the success path is going to be until we try two paths and see if one of them succeeds.

This is code we iterated to from something that was IMO less readable, because of much higher levels of indentation. See the discussion in the PR above, and https://github.com/google/go-github/pull/603#discussion_r109051657.

Also, this is consistent with our code at https://github.com/google/go-github/blob/d705ac853513f240ee84a8e4772b7126498f9591/github/repos_contents.go#L154-L167.